### DISCUSSION TITLE
BaseLine Update WorkFlow config

### DISCUSSION THREAD

**Discussion by c9mb** [umbraco/Umbraco.Cloud.Issues#1011] [Ideas]
We have multiple sites based on a baseline master, and plan to deploy many more sites this way.

The baseline master provides us with the initial default deployment of a site, and the child site is then customised to meet the client's specific requirements.

As the baseline is updated with code changes from both internal development and Umbraco upgrades, these are deployed to the child sites, and for the most part this works ok.

However, for some files, updating the baseline master to improve and/or enhance future default deployments, results in existing child site customisations being nuked, and additional work required to recover those customisations. Examples of such files might be css, js, fonts, images, etc.

You can't control this action using gitignore, as that would break the ability to update projects from local development, and so controlling this requires the ability to override the baseline update deployment workflow.

I would like to be able to specify such files in a .config, that the baseline update workflow examined as it deployed changes to existing child sites, so that changes to these files could be ignored as required. 

UmbracoBaseline.config
`<childupdates> 
      <ignore>
            <add path="/pathmatch/filematch*.css" />
            <add path="/pathmatch/filematch*.js" />
      </ignore>
</childupdates>`

The baseline update workflow could then read that .config file and process child updates using that as a control.

Being a .config could also offer the Umbraco devs team a tool to extend other control features people might want to introduce into the baseline workflow in the future.

Also, being a .config opens the door to transformations for environment-specific workflows, although I don't have any specific applications in mind for that at the moment.

**Comment by azure-devops-sync** [comment]
AB#642

**Comment by c9mb** [comment]
FWIW: I think that this essentially is a temporary change to the gitignore files on the child-sites being updated, prior to the baseline update running, and then undoing those changes to the child-site gitignore file afterwards. i.e. in the previous example, the baseline update would read the config file for processing child updates, modify the gitignore on the child-sites being updated to match the specs of the file, then run the baseline update, then undo the changes to the child-site gitignore files. Essentially, baseline child-update specific gitignore settings for child-sites.

**Comment by nul800sebastiaan** [comment]
Thanks @c9mbundy - just categorizing older issues on the tracker. This is a super cool idea and it's already given my colleague a headache thinking about all the moving parts involved! 😅 

At the moment we don't have the bandwidth to look at this soon. It is on our list of ideas to look at when we work on new baseline features though as it's a great idea! 👍

**Comment by c9mb** [comment]
I encountered another case today, where this would have been useful.

I needed to make some generic changes to some key doc-types (including compositions) on the baseline master, to deploy an enhanced configuration to new child-sites, but I absolutely cannot have these changes deployed to some of the existing child-sites via the baseline update workflow, as it will trash the data on those sites.

Instead, I need to still be able to push a customized version of those changes to those child-sites by deploying from the local dev-site of each site.

i.e. Deploy new changes to the baseline, and allow these to be rolled out when creating new child-sites. Enable protection from these baseline changes during baseline updates for some existing child-sites. Still allow deployment of them (duly customized) to protected child-sites from local dev-site, so I can make the required adjustments to ensure they are safe to deploy - or perhaps not deploy at all.

Which I think gets back to something along the lines of my OP - where there is a config file read by the baseline update workflow before it pushes changes to child-sites.

**UmbracoBaseline.config**
```
<childupdates>
    <ignore>
        // BaselineUpdate must not deploy changes to these files for this child-site
        // any changes to them must be deployed from the local dev-site
        <add path="/data/revision/document-type_blablabla.uda" />
        <add path="/css/site-styles.css" />
        <add path="/scripts/site-scripts.js" />
    </ignore>
</childupdates>
```

In the end I couldn't deploy the new changes to the baseline, as I can't protect the existing child-sites from them during a baseline update when pushing other baseline changes, but was forced to deploy a specific implementation of the required changes to each existing site from its local dev-site, and then create a list of these changes to be added post-deploy for new child-sites... which is rather negating the purpose of baseline deployment.

**Comment by sitereactor** [comment]
For the 'updated doc type' scenario could an approach be to create a new master for those new sites? This would introduce a second level for some sites, but you would have the separation for doc type.
Master (original) -> Child-Master (doc type change) -> New Baselines.
One caveat being that changing doc types in the original would push those changes through to Child-Master and New Baselines when and if you need to push changes all the way through.
You can create multiple levels of Baselines (a Child can also be a Master Project), but deployments work for one level: Master to Child.

**Comment by sitereactor** [comment]
I don't see the scenario of deploying only certain files or ignoring certain files from being deployed as being possible in the current setup.
All deployments are done through the git repository for each environment (in Master -> Live environment and Child -> Development or Live environment), which are connected as the Child having an upstream branch to the Master repository - very similar to how Forks work on Github. So it's all or nothing.

For customizations in Child projects we generally recommend to extend what originates from the Master. So if you have Styles.css in the Master and want to extend/change that, then create Styles.Site1.css with the changes instead of changing Styles.css. Having a clean cut of what originates from the Master and what is unique to a given Child project will go a long way in terms of having a smooth flow of updates for s baseline setup.

**Comment by c9mb** [comment]
I see, so there is no .gitignore used? 
Essentially, what I was describing was a dynamic .gitignore - the baseline master update workflow dynamically adjusting the child-site's .gitignore by the entries in the config file, before updating the child-site, and then putting it back to it's original state when done, so it doesn't affect deployments from other sources, including other environments.

**Comment by c9mb** [comment]
I guess that a .gitignore transformation is a better title for the config I was describing - one that only the baseline update workflow knows how to apply - but obviously of no use if the workflow doesn't honor a .gitignore in the update process.

**Comment by sitereactor** [comment]
The two git repositories (Master and Child) are merged when a deployment is started. The Child repository will do a fetch and a merge from the Master project. So anything in the repository will be merged into the Child project. The merge strategy will pick "development" files from the Master and config files from the Child in order to honor the flow.

Once the file is in the Master repository we can't really pick and choose what to deploy. At least it would be a completely different type of deployment than what we have today.
If you use gitignore it will only really be useful locally. Once it's in the repository it will be deployed, as this is essentially post-something-was-ignored-or-force-added.

If you can elaborate a little more on the concrete scenario it will help me understand what you want to achieve. The Doc Type change I understand, but the stylesheet/javascript ignore filter I'm not sure I understand. At least, right now, I'm inclined to say that following the recommendations around extending and not changing Master files would be the way to go.

**Comment by c9mb** [comment]
Perhaps this is moot now, as when I started this it was Mar-2019 and our v7 baseline was already heavily developed, so I was looking to avoid needing to rewrite some of the logic for handling things like child-site customizations. 

The baseline is a single universal master for all of our child sites, and uses a minimizing/bundling system to generate style & script files, and this is reflected in the child-site initial deployment. The child-site customizations are then applied by extending the baseline files and recompiling those bundles on the dev site, before deploying direct to the child site - the KISS approach.

But because of this approach, when a change to the baseline (to update some common styles or functions) is deployed to the existing child sites, it can break them, requiring that I then re-compile the child-site bundles on the local dev site and re-deploy the customized versions.  Not the end of the world for the number of sites we currently have, and typically only sees a site broken for 10mins or so as I recompile and redeploy, but would be an issue if I had 100's of child-sites hooked to the master.

The v7 baseline is the result of quite a few years work, and because of COVID-19 our plans to develop a v8 baseline to replace it have been delayed (we are going to struggle with the Feb-2021 cut-off) but perhaps it's time to start redesigning the approach for the v8 baseline, so this is not as much of an issue for us.

Of course, changing something like a doc-type (particularly a composition) or data-type (ugh!) will still be problematic, and I've been trying to avoid needing to maintain multiple baselines, but it may need a re-think given your description of the workflow.
